src/pybind/mgr/volumes/subvolume.py

```diff
@@ -0,0 +1,275 @@
+            # layout remains unset and will undesirably change with ancestor's
+            # pool layout changes.
+            xattr_key = 'ceph.dir.layout.pool'
+            xattr_val = self._get_ancestor_xattr(subvolpath, "ceph.dir.layout.pool")
+        # TODO: handle error...
+        self.fs.setxattr(subvolpath, xattr_key, xattr_val.encode('utf-8'), 0)
+
+    def remove_subvolume(self, spec):
+        """
+        Make a subvolume inaccessible to guests.  This function is idempotent.
+        This is the fast part of tearing down a subvolume: you must also later
+        call purge_subvolume, which is the slow part.
+
+        :param spec: subvolume path specification
+        :return: None
+        """
+
+        subvolpath = spec.fs_path
+        log.info("deleting subvolume with path: {0}".format(subvolpath))
+
+        # Create the trash directory if it doesn't already exist
+        trashdir = spec.trash_dir
+        self._mkdir_p(trashdir)
+
+        trashpath = spec.trash_path
+        self.fs.rename(subvolpath, trashpath)
+
+    def purge_subvolume(self, spec):
+        """
+        Finish clearing up a subvolume that was previously passed to delete_subvolume.  This
+        function is idempotent.
+        """
+
+        def rmtree(root_path):
+            log.debug("rmtree {0}".format(root_path))
+            try:
+                dir_handle = self.fs.opendir(root_path)
+            except cephfs.ObjectNotFound:
+                return
+            d = self.fs.readdir(dir_handle)
+            while d:
+                d_name = d.d_name.decode('utf-8')
+                if d_name not in [".", ".."]:
+                    # Do not use os.path.join because it is sensitive
+                    # to string encoding, we just pass through dnames
+                    # as byte arrays
+                    d_full = "{0}/{1}".format(root_path, d_name)
+                    if d.is_dir():
+                        rmtree(d_full)
+                    else:
+                        self.fs.unlink(d_full)
+
+                d = self.fs.readdir(dir_handle)
+            self.fs.closedir(dir_handle)
+            self.fs.rmdir(root_path)
+
+        trashpath = spec.trash_path
+        rmtree(trashpath)
+
+    def list_subvolumes(self, spec):
+        """
+        Returns a list of subvolumes within a subvolume group
+        """
+        grouppath = spec.group_path
+        return self._list_dir(grouppath)
+
+    def get_subvolume_path(self, spec):
+        path = spec.fs_path
+        try:
+            self.fs.stat(path)
+        except cephfs.ObjectNotFound:
+            return None
+        return path
+
+    ### Group operations...
+
+    def create_group(self, spec, mode=0o755):
+        path = spec.group_path
+        self._mkdir_p(path, mode)
+
+    def remove_group(self, spec):
+        path = spec.group_path
+        self.fs.rmdir(path)
+
+    def list_groups(self):
```

Comment:
Let's not introduce listing of subvolumes in this commit. It should be a separate commit. Last time I introduced listing, Patrick had certain issues around this,
#27594 (comment)

Reply:
ditto as per subvolume list.